Remove special case for choice rule IsPresent

kbrock opened the pull request:
presence_check is rolled back into true?()

For the non-present "IsPresent" case, looking up the lhs raises an exception
So we need to look up the predicate (rhs) first.
There is a test case to ensure people don't change this back

<!--
1. Describe what this Pull Request does and why you think it is needed.
   If this PR includes UI or CLI changes, please include Before/After screenshots
   If this PR includes performance changes, please include Before/After metrics showing improvement.
-->

<!--
2. If this fixes an existing issue, please specify in `Fixes #<id>` format
   (As described in https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)
-->

<!--
3. Ask @miq-bot to apply a scope label (bug, enhancement, etc) and any additional reviewers or assignees.
   (As described in https://github.com/ManageIQ/miq_bot#requested-tasks)
   e.g. `@miq-bot add-label label_name`
        `@miq-bot add-reviewer @name`
-->

kbrock commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -23,42 +23,39 @@ def initialize(_workflow, _name, payload)
+          predicate
         end
 
+        # Check if a value is numeric
+        # predicate is whether the type matches: {"Variable": "$.foo", "IsNumeric": true} i.e.: true
+        # @param value The value to check
+        # @param predicate [Boolean] Whether we expect the value to be numeric (true) or not present(false)
+        # @return [Boolean] Whether the value matches the expectation
```
I feel the way `predicate` works with the check is confusing.
Do we need to document this better?
Or do people understand it just fine - and I already documented this well enough?

Summary
=======

```ruby
def is_null?(value, predicate)
  if value.nil? # it was nil - check passed
    if predicate # was hoping that it is nil (and it was nil)
      true
    else # was hoping that it was not nil (and it was nil)
      false
    end
  else # it was not nil
    if predicate # was hoping that it is nil (and it was not nil)
      false
    else # was hoping that it was not nil (and it was not nil)
      true
    end
  end
end

# condense to:
def is_null?(value, predicate)
  value.nil? == predicate
end
```

kbrock replied:
put this into the code so hopefully I can stop feeling the need to explain this.

Fryguy commented:
Can you explain what lhs and rhs are with an example? I find this statement very confusing: "For the non-present "IsPresent" case, looking up the lhs raises an exception".  Maybe the concept of left and right are what don't make sense to me.

kbrock commented:
@Fryguy lol. so apparently my concern over this being confusing is founded.

Yea, Json gives us 2 variables. `"Variable": $lhs` and `"IsPresent": $rhs`. For comparison operations, `lhs` and `rhs` makes sense. Unfortunatly, Type checks (i.e.: `"IsNull"`) and to some extent `"StringMatch"` the nomenclature leaves something to be wanted. So the methods use variable `value` and `predicate` where `lhs=value` and `rhs=predicate`.

LMK if the added comments make this more understandable. Can't remember where the predicate word originated. It is now in method names and may be to late to rename to `desired_vale` or `expected_value`

So much for `+6/-14`.

Update:
- Added longer explanation for how we came up with `value.nil? == predicate` in the code.
- Added examples for the `true?()` for type check and comparisons
- Added comments on logic methods

kbrock commented:
update:

- changed pseudo code to a truth table
- minor wording change in exception case (the value is not present for IsPresent)

kbrock commented:
update:
- rebased (just to make history linear)

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -22,55 +22,120 @@ def initialize(_workflow, _name, payload)
-          return presence_check(context, input) if compare_key == "IsPresent"
-
-          lhs = variable_value(context, input)
+          # right hand side of comparison or predicate if this a type check
           rhs = compare_value(context, input)
+          # left hand side of comparison or value of the variable if this is a type check
+          lhs = variable_value(context, input)
+          # Example:
+          #
+          # {"Variable": "$.foo", "IsNumeric": true}
+          #   value     = lhs = input["$.foo"]
+          #   predicate = rhs = true
+          #               true meaning we expect $.foo to be numeric.
+          # => is_numeric?(value, predicate)
+          #
+          # {"Variable": "$.foo", "GreaterThanString": "aaa"}
+          #   lhs = input["$.foo"]
+          #   rhs = "aaa"
+          # => op_greaterthan?(lhs, rhs)
+          #
+          # {"Variable": "$.foo", "GreaterThanNumericPath": "$.bar"}
+          #   lhs = input["$.foo"]
+          #   rhs = input["$.bar"]
+          # => op_greaterthan?(lhs, rhs)
+          #
```
I think you can simplify this comment to 


```suggestion
          # Payload pattern is: {"Variable": $lhs, $operator: $rhs}
          # Example:
          #
          #   {"Variable": "$.foo", "IsNumeric": true}
          #     lhs = input["$.foo"]
          #     rhs = true
          #     is_numeric?(lhs, rhs)
          #
          #   {"Variable": "$.foo", "GreaterThanString": "aaa"}
          #     lhs = input["$.foo"]
          #     rhs = "aaa"
          #     op_greaterthan?(lhs, rhs)
          #
          #   {"Variable": "$.foo", "GreaterThanNumericPath": "$.bar"}
          #     lhs = input["$.foo"]
          #     rhs = input["$.bar"]
          #     op_greaterthan?(lhs, rhs)
          #
          # NOTE: IsPresent works a little differently as the lhs might raise a PathError. See the exaception handler below.
          rhs = compare_value(context, input)
          lhs = variable_value(context, input)
```
